src/kaleidoscope_internal/typed_plugins.h

```diff
@@ -0,0 +1,226 @@
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2018  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+#include "kaleidoscope/macro_map.h"
+
+// The following is template code that generates an array of pointers of
+// a specific type of plugins (e.g. LEDMode's) at compiletime and can be
+// stored in PROGMEM.
+
+namespace kaleidoscope {
+namespace internal {
+namespace typed_plugins {
+
+// The TypedPluginArray class stores pointers to plugins of a specific
+// type in a recursive fashion.
+//
+template<typename TargetType__,
+         bool IsAppropriateType__,
+         bool...MoreTypeInfo__>
+class TypedPluginArray {
+
+ public:
+
+  typedef TypedPluginArray<TargetType__, MoreTypeInfo__...> NestedArray;
+
+  template<typename...MorePlugins__>
+  constexpr TypedPluginArray(TargetType__ *entry, MorePlugins__...more_plugins)
+    : entry_(entry),
+      nested_array_(more_plugins...) {}
+
+  static constexpr uint8_t n_entries
+    = NestedArray::n_entries + 1;
+
+  typedef TargetType__ ContentType;
+
+ private:
+  TargetType__ *entry_;
+  NestedArray nested_array_;
+} __attribute__((packed)); // Make sure that there are no padding
+// bytes added by the compiler.
+// This is important to let the class
+// have the same layout as a POD array.
+
+
+template<typename TargetType__,
+         bool...MoreTypeInfo__>
+class TypedPluginArray<TargetType__,
+        false /* Not of appropriate type */,
+        MoreTypeInfo__...> {
+
+ public:
+
+  typedef TypedPluginArray<TargetType__, MoreTypeInfo__...> NestedArray;
+
+  template<typename...MorePlugins__>
+  constexpr TypedPluginArray(void */* non-matching plugin */,
+                             MorePlugins__...more_plugins)
+    : nested_array_(more_plugins...) {}
+
+  static constexpr uint8_t n_entries
+    = NestedArray::n_entries;
+
+  typedef TargetType__ ContentType;
+
+ private:
+
+  NestedArray nested_array_;
+} __attribute__((packed));
+
+template<typename TargetType__>
+struct TypedPluginArray<TargetType__, true /* is of appropriate type */> {
```

Comment:
Am I correct in guessing that this template is overriding the definition above, for the specific case where the second parameter is `true`? Also, is it important (or relevant) that this definition uses the `struct` keyword, whereas the previous template definitions of `TypedPluginArray` use the `class` keyword?

Reply:
> Am I correct in guessing that this template is overriding the definition above, for the specific case where the second parameter is true?

Yes, correct.

> Also, is it important (or relevant) that this definition uses the struct keyword, whereas the previous template definitions of TypedPluginArray use the class keyword?

There is no restriction on specializations about class/struct with respect to the general implementation. As a struct is a class with default public access, this makes sense. In general I tend to use struct for those classes that don't have private/protected members to save the `public:`.

So you are right, in line 88 it should read `class` and the `public:` in the `struct` body at line 107 is superfluous.